[1단계 - 출석 구현] 제프(변우영) 미션 제출합니다.

horizonpioneer opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [v] 미션의 필수 요구사항을 모두 구현했나요?
- [v] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [v] 애플리케이션이 정상적으로 실행되나요?
- [v] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. --> https://prolog.techcourse.co.kr/studylogs/3949

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
getter 및 setter를 사용했고, edit과 같이 메서드나 상수 이름에 축약형을 사용했기 때문입니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

<페어와 미션을 진행하면서 고민했던 부분>
**1. TDD 방식으로 설계 시, 클래스 구현에 대한 고민**
TDD 방식으로 설계를 할 때, 임시 클래스 안에서 테스트에 필요한 기능들을 구현하다가 마지막 리팩토링 단계에서 각각의 역할에 맞는 객체를 만들고 책임을 분리하는게 좋을지,  특정 도메인 객체를 만들면서 진행하는게 좋을지 고민이 되었습니다. 이번 페어 미션에서는 전자의 방식으로 진행을 했지만, TDD를 진행하면서 특정 객체의 역할을 주어야 겠다는 생각이 들어도 일단 임시 클래스로 쭉 구현을 하다가 마지막 단계에서 책임 분리를 하는 것이 좀 더 TDD의 본질적인 목적에 맞을지 이프의 생각을 듣고 싶습니다.


**2. 수정 기능이 필요한 객체 리스트를 특정 객체의 내부 필드로 사용하는 것과 일급컬렉션으로 직성하여 사용하는 것에 대한 고민**
 해당 미션에서 Crew라는 도메인이 List를 필드로 가지고 출석 기록을 추가/수정하는 방식으로 구현했습니다. 이 과정에서 Attendance의 일급 컬렉션(Attendances)을 따로 두고 이를 Crew의 필드로 둘지,아니면 List를 직접 필드로 둘지에 대한 고민이 있었습니다.

불변성을 지키기 어려운 경우, 즉 비즈니스 요구에 따라 수정이 필요한 객체 리스트의 경우에 일급 컬렉션으로 설정하는 것보다 특정 객체의 필드로 두는게 나을까요? 일반적으로 객체 리스트만 필드로 가진 객체를 생성하게 되면 다른 사람이 코드를 보았을 때 일급 컬렉션이라고 생각을 할 것이고, 그렇게 되면 불변성을 보장하는 객체라고 생각할 것 같습니다. 개인적으론 개발자간의 협업에 있어서 특정 의미로 굳어진 코드 방식은 오해의 소지가 없게 사용하는게 맞을 것 같다는 생각이 듭니다. 이프의 의견을 듣고 싶습니다!

**3. Day 객체의 필요성에 대한 고민과 공휴일 처리에 대한 고민**
처음 Day 객체를 구성했을 떄와 달리, PR 시점 기준으론 LocalDate 필드 하나만 가지게 되었고 인텔리제이 내부에서도 record 타입으로 변경을 하라는 추천을 받았습니다. Day 객체를 도입한 주된 이유는 주말을 제외한 공휴일 처리를 위한 것이었으며, 일반적인 방법으로 생각했을 때 공휴일을 직접 관리하기 위해 Day 객체를 활용하는 것이 적절하다고 판단했습니다. 

하지만, 현재 미션에서 사용하는 CSV 파일은 2월을 기준으로 작성되었고, 2월에는 공휴일이 없기 때문에 사실상 Day 객체가 하는 역할이 거의 없습니다. 추가적으로 날짜를 Day라는 커스텀 클래스 객체로 관리하게 되면서 도메인간의 계층 구조가 더욱 복잡해진 것도 사실입니다.

이러한 상황에서 향후 확장을 고려하여 Day 객체를 그대로 유지할지, 현재 필요하지 않으므로 Day 객체를 제거하고 LocalDate를 직접 사용할지에 대한 고민이 있습니다. 또한, 현재 주말을 제외한 공휴일은 해당 월 특성에 따라 구현하지 않았지만, 특정 공휴일과 관련된 라이브러리 없이 주어진 날짜 객체들만 가지고 공휴일 처리를 따로 할 방법이 있을까요? 이프의 생각이 궁금합니다!

**4. 생성자 내부에 특정 로직을 넣어도 괜찮을까요**
초기화를 위한 특정 로직과 검증 로직 여러개를 넣어 구현해도 괜찮을까요? 관련 내용에 대해 알아보다 생성자에 여러 코드가 존재할 때, 성능 최적화 이슈가 있다는 내용을 보게되었는데 리소스가 낭비되는 정도의 차이가 클까요? Attendance 생성자 내부 코드에 updateStatus 메서드를 통해 특정 조건에 따라 필드값이 변경되도록 구현했는데 다른 단점도 존재하는지 이프의 의견을 듣고 싶습니다!

horizonpioneer committed:
docs(readme): 1차 요구사항 정리

horizonpioneer committed:
test: 출석 확인 정상 출력 테스트 구현

horizonpioneer committed:
feat(attendanceCheck): 등교 시간에 따라 출석 상태를 체크하는 기능 구현

horizonpioneer committed:
test: 등교 시간에 따른 출석 상태 변동 테스트 구현

horizonpioneer committed:
feat(attendanceCheck): 날짜에 따라 해당 요일을 반환하는 기능 구현

horizonpioneer committed:
test: 출석 확인 시, 현재 날짜가 출력되는 테스트 구현

horizonpioneer committed:
feat(AttendanceCheck): 문자열을 LocalTime으로 변환하는 기능 구현

horizonpioneer committed:
test: 문자열을 LocalTime으로 변환하는 테스트 구현

horizonpioneer committed:
feat(AttendanceCheck): 날짜에 따른 등교 기준 시간을 반환하는 기능 구현

horizonpioneer committed:
test: 날짜에 따른 출석 기준 시간 적용 테스트 구현

horizonpioneer committed:
refactor: InputConverter 유틸 클래스 구현

- 문자열을 LocalTime으로 변환하는 메서드를 분리

horizonpioneer committed:
refactor: 요일 및 출석 시간 enum화한 DayOfWeek 구현

- id를 기반으로 요일 정보를 반환하는 정적 메서드 구현

horizonpioneer committed:
refactor: Day와 Attendance 도메인 분리

- 출석 상태를 확인하는 기능 분리

horizonpioneer committed:
refactor: AttendanceDto 구현

horizonpioneer committed:
refactor: LocalDate를 기반으로 DayOfWeek 객체를 반환하는 기능 구현

horizonpioneer committed:
refactor: Dto 기반으로 테스트 수정

horizonpioneer committed:
test: 수정시간에 따라 기존 출석시간 변경 테스트 구현

horizonpioneer committed:
feat: AttendanceTime 변경 메서드 및 Dto 필드 추가

- setAttendanceTime을 통해 출석 시간을 변경하고 updateStatus를 호출하여 출석 상태를 변경할 수 있도록 한다.

horizonpioneer committed:
refactor: 여러 개의 출석 상태 변경 테스트를 ParameterizedTest로 통합

horizonpioneer committed:
test: 각 출석 상태별 횟수 계산 테스트 구현

horizonpioneer committed:
test: 제적 위험자 여부 판단 테스트 구현

- PenaltyStatus enum을 통해 제적 위험자 기준에 따라 StatusName 반환 기능 추가

horizonpioneer committed:
refactor: 공휴일 여부를 나타내는 필드를 메서드에서 산출하도록 변경

- isHoliday 필드 삭제 후, checkHoliday 메서드를 통해 공휴일 여부 판단

horizonpioneer committed:
refactor: 제적 상태 이름을 반환하는 기능을 제적 상태 enum을 반환하는 기능으로 수정

horizonpioneer committed:
feat: Crew 도메인 추가

horizonpioneer committed:
feat: CrewDto 클래스 구현

horizonpioneer committed:
feat: CrewDto의 중첩 Dto CrewDtos 구현

horizonpioneer committed:
env: 출석 기록 파일 추가

horizonpioneer committed:
feat: 출석 기록으로부터 크루원 초기 데이터 세팅

- 출석 기록을 csv 파일에서 가져오는 기능 구현

horizonpioneer committed:
feat: LocalTime을 String으로 변환하는 기능 구현

horizonpioneer committed:
feat: 기존 출석 기록이 있는지 확인하는 메서드 구현

horizonpioneer committed:
fix: csv 파일 경로 수정

horizonpioneer committed:
feat: InputView 클래스 구현

horizonpioneer committed:
feat: OutputView 클래스 구현

horizonpioneer committed:
feat: 닉네임을 통해 Crew를 반환하는 메서드 구현

horizonpioneer committed:
refactor: AttendanceDto 멤버로 출석 일자 추가

horizonpioneer committed:
feat: 문자열을 정수로 바꾸는 정적 메서드 구현

horizonpioneer committed:
chore: 상수명 변경

horizonpioneer committed:
feat: 일자에 따라 Attendance 객체를 반환하는 기능 구현

horizonpioneer committed:
feat: Crew의 출석 기록을 반환하는 기능 구현

- 날짜 기반 정렬 후, 방어적 복사를 이용하여 getter 구현

horizonpioneer committed:
feat: 출석하지 않은 날짜를 일괄적으로 결석 처리하는 기능 구현

horizonpioneer committed:
refactor: 분리된 도메인으로 테스트 재구성

- 임시 클래스인 AttendanceCheck 삭제 후, 테스트 수정

horizonpioneer committed:
feat: OutputView 클래스 구현

horizonpioneer committed:
refactor: covertLocalTimeToString 메서드에 null 처리 추가

horizonpioneer committed:
refactor: 공휴일에 Attendance 객체가 생성되는 경우, 예외 처리 추가

horizonpioneer committed:
refactor: crewDtos 정렬 방식 수정

horizonpioneer committed:
refactor: 출석 상태 확인 메서드에 null 처리 구현

horizonpioneer committed:
feat: CrewDto의 리스트 생성 메서드 구현

horizonpioneer committed:
refactor: AttendanceHistoryLoader 메서드 분리

horizonpioneer committed:
feat: AttendanceController 구현

horizonpioneer committed:
refactor: AttendanceController 메서드 분리

horizonpioneer committed:
feat: Application 추가

horizonpioneer committed:
refactor: Day 멤버 final로 선언

horizonpioneer committed:
chore: getStandardTimeById 메서드 삭제

horizonpioneer committed:
refactor: DayofWeek 필드 제거

- 내부 메서드를 통해 DayOfWeek 조회하도록 수정

horizonpioneer committed:
refactor: OutputView 메서드 정리

horizonpioneer committed:
refactor: 상수 text block 처리 및 사용하지 않는 상수 제거

horizonpioneer committed:
refactor: 재할당이 필요하지 않은 필드에 final 추가

horizonpioneer committed:
refactor: 테스트 코드 내부 필드에 final 추가 및 사용하지 않는 필드 제거

horizonpioneer committed:
refactor: 가독성을 위해 코드 정리

horizonpioneer committed:
refactor: 출석 기록 로드 과정에서 예외 발생 시, 콘솔에 StackTrace를 남기는 대신 예외가 발생하도록 수정

sinb57 commented on src/main/java/config/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,60 @@
+    private final DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+    private final DateTimeFormatter timeFormatter = DateTimeFormatter.ofPattern("HH:mm");
+
+    public Crews loadCrews() {
```
해당 로직은 테스트가 필요없을까요?

horizonpioneer replied:
테스트 코드 추가해보겠습니다!

sinb57 commented on src/main/java/config/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,60 @@
+package config;
```
config 패키지는 어떠한 역할을 지니고 있나요?

horizonpioneer replied:
파일에서 특정 정보를 읽어오는 역할에 대해 고민하다가 util 패키지와 config 패키지로 의견이 갈렸었는데 해당 부분에 대해 깊게 토론하지 않고 넘어간 것 같습니다! 앱 설정 파일을 읽고 파싱하는 클래스는 config 패키지에 들어갈 수 있겠다고 생각하고 넘어갔었는데, 다시 생각해보니 해당 파일 로더는 출석 기록을 가져와 Crews를 만드는데 영향을 주기 때문에 domain 패키지에 들어갈 수 있을 것 같습니다.

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,108 @@
+        processAttendance(option);
+        processAttendanceEdit(option);
+        processAttendanceHistory(option);
+        processPenaltyCheck(option);
```
if 문의 나열과 다르지 않네요. 그렇기에 오히려 if-else보다 성능적으로 좋지 않아요.
else 예약어를 사용하지 말라는 제약이 왜 주어졌을까요? 어떻게 하면 if-else 없이 분기로직을 처리할 수 있을지 고민해보고 적용해보세요. 객체지향의 다형성을 적극 활용해보시기 바랍니다~

horizonpioneer replied:
감사합니다! 해당 부분 좀 더 고민해보고 수정하겠습니다!

sinb57 commented on README.md:
```diff
@@ -1,3 +1,142 @@
+
+- [ ]  실행할 기능을 숫자로 입력 받는다
+    - [ ]  1 → 출석 확인
+    - [ ]  2 → 출석 수정
```
요구사항을 충족했다면 체크 부탁드려요~

horizonpioneer replied:
살아있는 기능 목록을 만들었어야 했는데 TDD를 진행하면서 문서 업데이트를 깜빡한 것 같습니다! 업데이트해보겠습니다

sinb57 commented on src/main/java/config/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,60 @@
+            String date = datetimeValues[0];
+            String time = datetimeValues[1];
+
+            Crew crew = getCrew(crewMap, crews, values[0]);
```
메서드를 분리하면 의미 전달이 뚜렷해질 듯 해요

horizonpioneer replied:
감사합니다! 코드 의미 전달에 더 신경써서 리팩토링 해보겠습니다!

sinb57 commented on src/main/java/config/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,60 @@
+        Map<String, Crew> crewMap = new HashMap<>();
+
+        try (BufferedReader reader = new BufferedReader(new FileReader("src/main/resources/attendances.csv"))) {
+            reader.readLine();
```
어째서 첫줄을 무시하는지 의미부여하면 좋겠어요.

sinb57 commented on src/main/java/config/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,60 @@
+
+    public Crews loadCrews() {
+        List<Crew> crews = new ArrayList<>();
+        Map<String, Crew> crewMap = new HashMap<>();
```
crew를 등록하는 과정을 Crews 에 위임하면 Loader 로직가 가벼워질 듯 해요.

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,108 @@
+    public void processPenaltyCheck(String option) {
+        if (!option.equals("4")) return;
+
+        List<CrewDto> crewDtos = crews.createCrewDtos();
+        List<CrewDto> penaltyCrewDtos = crewDtos.stream()
+                .filter(crewDto -> crewDto.getPenaltyStatus() != PenaltyStatus.NONE)
+                .collect(Collectors.toCollection(ArrayList::new));
+
+        outputView.printPenaltyCrews(new CrewDtos(penaltyCrewDtos));
```
CrewDtos 가 필요한가요? penaltyCrewDtos 를 outputView에 전달하면 안되나요??

horizonpioneer replied:
조건에 따라 정렬하는 로직을 Dto 내부에 두었었는데, OutputView로 책임을 옮기고 penaltyCrewDtos를 전달 받아 사용하게 하면 Dtos는 사용하지 않아도 될 것 같습니다!

sinb57 commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,56 @@
+            isAbsent = true;
+            return;
+        }
+        isLate = isLate();
+        isAbsent = isAbsent();
+    }
+
+    private boolean isLate() {
+        return day.isLate(attendanceTime);
+    }
+
+    private boolean isAbsent() {
+        return day.isAbsent(attendanceTime);
+    }
```
메서드를 분리하여 얻은 장점이 궁금합니다.

horizonpioneer replied:
클래스 밖에서도 사용하기 위해 만들었는데 private으로 제한해버려서 의미가 없는 메서드가 되어버렸네요...

sinb57 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    private Integer calculateNonAttendanceCount() {
+        return calculateAbsentCount() + calculateLateCount() / 3;
```
매직넘버가 보이네요

sinb57 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,99 @@
+
+    private Integer calculateAbsentCount() {
+        return (int) attendances.stream()
+                .filter(attendance -> attendance.toDto().getAbsent().equals(true))
```
왜 Dto로 변환하여 getter를 통해 값을 가져다 비교하는지 궁금합니다.
`attendance.isAbsent()`하면 되지 않을까요?
제프가 생각하는 DTO의 역할은 무엇인가요??

horizonpioneer replied:
Attendance에서 isLate와 isAbsent 메서드를 private으로 제한해버리는 바람에 해당 메서드의 의미가 사라지고 밖에선 getter를 사용해 값을 비교하는 꼴이 되어버렸네요... 
DTO의 역할은 계층 간 데이터 전달이며, 객체의 상태를 비교하거나 외부 값과 상호작용하는 작업은 도메인 내부에서 수행하는 것이 더 객체지향적인 설계에 부합한다고 생각합니다.

sinb57 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,99 @@
+            }
+        }
+        return false;
+    }
```
> indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.

요구사항을 지켜주세요~

horizonpioneer replied:
stream을 활용해 수정해보겠습니다. 감사합니다!

sinb57 commented on src/main/java/util/Converter.java:
```diff
@@ -0,0 +1,39 @@
+
+    public static LocalTime convertStringToLocalTime(String timeInput) {
+        try {
+            DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm");
```
메서드가 호출될 때마다 객체가 생성되는데 재사용할 수 없을까요?

horizonpioneer replied:
formatter를 상수로 빼면 될 것 같습니다!

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,108 @@
+        if (!option.equals("2")) return;
+
+        Crew crew = crews.findByNickname(inputView.getEditNickname());
+        Attendance attendance = crew.findByDate(Converter.convertStringToInteger(inputView.getEditDayOfMonth()));
```
inputView으로부터 String 값을 전달 받아야만 하나요?

horizonpioneer replied:
InputView에서 Converter를 호출해 int 값으로 바로 변환 받아서 사용하는게 나을까요?

sinb57 replied:
InputView에서 값을 가공하여 반환해도 문제 없다고 봐요. "입력"이라는 행위는 무엇일까요? 콘솔로 문자열을 입력받는 것만을 의미할까요? 값을 제공받는 입장에서는 받아서 사용하기 편한 값으로 넘겨받으면 좋지 않을까요?

sinb57 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,118 @@
+    private static final String LATE_COUNT = "지각: %d회\n";
+    private static final String ABSENT_COUNT = "결석: %d회\n\n";
+    private static final String PENALTY_CREW_READ_RESULT_PREFIX = "제적 위험자 조회 결과";
+    private static final String PENALTY_CREW_READ_RESULT = "- %s: 결석 %d회, 지각 %d회 (%s)\n";
```
문자열을 재사용하지 않음에도 상수로 분리한 이유가 무엇인가요?

horizonpioneer replied:
출력해야 하는 문자열의 목록을 한 눈에 볼 수 있으면 수정에도 용이하다고 생각해서 따로 상수 처리했는데, 재사용하지 않는 특정 문자열은 문자열 그대로 메서드에 적는 것이 좋을까요?

sinb57 replied:
수정이 빈번하다면 편의를 위해 제프의 방법대로 상수 처리하기도 해요. 하지만 수정이 빈번하지 않다면 오히려 불편해집니다. 코드를 읽는 시간과 경우가 많다면, 상수를 들여다보지 않고서도 빠르게 이해할 수 있어야 하는데요. 상수명으로는 명확한 의도 파악이 어렵습니다.

아래의 코드를 봤을 때, 어떤 코드가 읽고 파악하기 수월하신가요? 저는 포맷팅에 맞춰 바로 파악할 수 있는 2번째가 좋아 보이네요.
```java
        System.out.printf(UPDATE_COMPLETE_MESSAGE,
                date.getMonth().getValue(), date.getDayOfMonth(), dayOfWeekName,
                originalAttendanceTime, originAttendanceStatusName,
                editedAttendanceTime, editedAttendanceStatusName);
```

```java
        System.out.printf("%d월 %02d일 %s %s (%s) -> %s (%s) 수정 완료!\n",
                date.getMonth().getValue(), date.getDayOfMonth(), dayOfWeekName,
                originalAttendanceTime, originAttendanceStatusName,
                editedAttendanceTime, editedAttendanceStatusName);
```

horizonpioneer replied:
재사용하지 않는 문자열이라면 해당 출력 메서드에 문자열 그대로 적는 것이 의도 파악하기 쉬울 것 같습니다!
밑의 코드가 더 가독성이 좋아보입니다.

horizonpioneer replied:
재사용하지 않는 문자열이라면 해당 출력 메서드에 문자열 그대로 적는 것이 의도를 파악하기 쉬울 것 같습니다!
밑의 코드가 더 가독성이 좋아보입니다.

sinb57 commented on src/main/java/domain/CrewDto.java:
```diff
@@ -0,0 +1,38 @@
+package domain;
+
+public class CrewDto {
```
DTO를 왜 만들었는지 궁금해요.
DTO는 무엇이며, 지금 어떠한 목적으로 사용하고 있나요?

horizonpioneer replied:
출력 조건에 따라 특정 양식을 출력하기 위한 데이터 전달 목적으로 DTO를 사용했습니다. 필요한 데이터를 묶어서 관리하기 위함이었는데, 계층 간 데이터 전달 목적이더라도 지양하는게 좋을까요?

sinb57 replied:
Data Transfer Object 이름 그대로 데이터 전달 목적으로 사용하긴 하지만, 현재 DTO의 사용방법이 좋지 않아보여요. 어째서 "계층간"이라고 용도를 특정지어 생각하셨나요? 제프는 도메인을 무엇이라 생각하시나요? 어째서 도메인과 view를 나눴다고 보시나요? 만약 출력 조건이 변경되어 DTO의 내용에 변화가 발생하면, 변경 범위를 어떻게 보시나요? 출력의 변경에 도메인에 변경이 이뤄지는 것이 옳을까요?

horizonpioneer replied:
계층 간 전달 목적이라고 생각한 이유는 도메인 내부 필드가 되도록 밖으로 노출되면 안된다고 생각했기 때문입니다. 출력과 같이 필요한 부분에 있어서 필요한 데이터만 주고 받는데 사용해야 한다고 생각했습니다. 만약 출력 조건이 변경된다면 DTO에 들어가야 할 필드 타입을 변경해야 한다고 생각합니다. 
다만 도메인에서 Dto로 특정 데이터를 담는 과정을 컨트롤러나 서비스 계층에서 구현할까 생각도 했었지만 해당 부분 피드백 받고 다시 생각해보고싶어서 도메인 내부에서 toDto 메서드를 사용해 변환하도록 했습니다. 
말씀해주신 도메인의 변경이라는 부분에 있어서 도메인 내부에 있는 toDto의 수정도 포함될까요? 도메인이 Dto를 알고 있어도 되나 하는 생각도 들었지만 아직 답을 찾진 못했습니다.

sinb57 replied:
> 계층 간 전달 목적이라고 생각한 이유는 도메인 내부 필드가 되도록 밖으로 노출되면 안된다고 생각했기 때문입니다. 출력과 같이 필요한 부분에 있어서 필요한 데이터만 주고 받는데 사용해야 한다고 생각했습니다. 만약 출력 조건이 변경된다면 DTO에 들어가야 할 필드 타입을 변경해야 한다고 생각합니다.

DTO를 처음 언급한 마틴 파울러는, 데이터 전송 비용을 아끼기 위해 전송하고자 하는 데이터를 묶고자 DTO라는 단어를 처음 말했어요. DTO를 "계층 간"에 국한하여 생각하지 않으셨으면 좋겠어요. 실제로 DTO는, 비즈니스 로직 간에 데이터를 전달하는 과정에서도 많이 사용해요.

> 다만 도메인에서 Dto로 특정 데이터를 담는 과정을 컨트롤러나 서비스 계층에서 구현할까 생각도 했었지만 해당 부분 피드백 받고 다시 생각해보고싶어서 도메인 내부에서 toDto 메서드를 사용해 변환하도록 했습니다.
말씀해주신 도메인의 변경이라는 부분에 있어서 도메인 내부에 있는 toDto의 수정도 포함될까요? 도메인이 Dto를 알고 있어도 되나 하는 생각도 들었지만 아직 답을 찾진 못했습니다.

여러 고민과 질문이 보이는데, 핵심은 하나예요. 도메인이 DTO에 의존하면 안되는 이유는 무엇일까요? 정확히는, 지금처럼 출력을 위한 DTO를 도메인이 의존하면 무슨 문제가 있을까요?

horizonpioneer replied:
Dto 내부 필드에 수정 사항이 생겼을 때 도메인의 코드까지 변경해야 하는 문제점이 있을 것 같습니다. 또한, Dto 내부 필드 중 참조형 객체가 있다면, 참조값을 그대로 넘겨서 원본 데이터에 영향을 줄 수 있다는 문제점도 있겠네요....
도메인은 Dto를 모르도록 분리시켜야 할 것 같은데 다른 계층에서 변환 과정을 구현하려면 결국 도메인의 getter를 활용할 수 밖에 없는 걸까요? Crew 기준으로 lateCount와 absentCount는 계산하는 메서드를 public으로 전환해 사용하면 해당 값을 밖에 전달할 수 있을 것 같은데 nickname과 같은 경우엔 getter를 만들 수 밖에 없겠네요

sinb57 replied:
아주 좋은 고민이에요ㅎ 사실 DTO는 Level2에서 깊게 고민하기 시작할 부분이고, Level1에서는 사용할 이유가 없어요. 제프는 왜 getter의 사용을 지양해야 한다고 생각하시나요? view에서 도메인에 의존하면 안되나요?

horizonpioneer replied:
아무 의미 없이 단순히 필드만 반환하는 것을 지양해야할 것 같습니다. getter를 통해 값을 가져와서 비즈니스 로직을 구현하는 부분을 지양해야 하는 것이고 출력을 위해 필요한 부분은 사용하는게 맞다고 생각합니다

sinb57 reviewed (changes requested):
안녕하세요 제프~ 미션 진행하시느라 고생 많으셨어요!!
질문에 대한 답변과 코드에 대한 코멘트 확인 부탁드려요~

공통 피드백) 
열려있는 모든 메서드에 대해 최대한 단위 테스트를 작성해주시고,
아래의 프로그래밍 요구사항에 유의하여 수정해주세요.
- indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.
- 함수(또는 메서드)의 길이가 10라인을 넘어가지 않도록 구현한다.
- else 예약어를 쓰지 않는다.

질문)
> 선택한 점수의 이유를 적어주세요.
> getter 및 setter를 사용했고, edit과 같이 메서드나 상수 이름에 축약형을 사용했기 때문입니다.

setter를 사용하셨다고 했는데, 어디에 있나요??

---

남겨주신 질문에 답변 드려요~

> 1. TDD 방식으로 설계 시, 클래스 구현에 대한 고민

사람마다 스타일이 다르겠지만, 저는 후자의 방식을 선호해요. 요구사항에 기반하여 대략적인 설계를 먼저 구상하고서 접근하다보니, 어떠한 객체들이 필요하고 어떠한 책임과 역할을 부여하면 좋을지 판단이 되거든요. 이렇게 기초를 쌓은 상태에서 테스트코드와 함께 구현에 대한 즉각적인 피드백을 주고받으며 개발하곤 합니다.

제프가 생각하는 테스트코드의 장점과 역할은 무엇인가요?

> 2. 수정 기능이 필요한 객체 리스트를 특정 객체의 내부 필드로 사용하는 것과 일급컬렉션으로 직성하여 사용하는 것에 대한 고민

List<Attendance>에 행위를 부여하여 응집력을 높이고 역할을 부여할 수 있다면, 충분히 Attendances를 만들 이유가 될텐데요. 제프는 왜 일급컬렉션이 불변을 보장해야 한다고 보시나요??

> 3. Day 객체의 필요성에 대한 고민과 공휴일 처리에 대한 고민

공휴일을 담지 않는, 즉 "주중"의 의미를 지닌 객체로 역할을 부여하면 어떨까 싶어요. 공휴일로 객체를 생성할 수 없도록 하면 되지 않을까요? 실제로 현재 `Day#checkHoliday` 메서드가 공휴일 검증에만 사용되는 것을 볼 수 있고, 요구사항도 공휴일에 대한 제약이 있다보니 Day 객체의 역할을 구체적으로 바꾸어봐도 좋아보여요.

> 4. 생성자 내부에 특정 로직을 넣어도 괜찮을까요

생성자에 검증 로직을 두는 것은 옳다고 생각해요. 해당 객체의 생성에 대한 계약 조건이므로, 생성 시점에 validate하는 것은 지극히 당연하다고 봐요. 다만 로직에 따라 값을 달리 부여하거나 하는 행위는 음 정적 팩토리 메서드를 활용하는데요. 정적 팩토리 메서드에 대해 알아보시고, 생성자와 어떠한 차이가 있는지 그리고 제프는 어떠한 기준으로 생성자와 정적 팩토리 메서드를 나누어 사용할지 한번 고민해보시면 좋겠어요

horizonpioneer commented:
> 안녕하세요 제프~ 미션 진행하시느라 고생 많으셨어요!! 질문에 대한 답변과 코드에 대한 코멘트 확인 부탁드려요~
> 
> 공통 피드백) 열려있는 모든 메서드에 대해 최대한 단위 테스트를 작성해주시고, 아래의 프로그래밍 요구사항에 유의하여 수정해주세요.
> 
> * indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.
> * 함수(또는 메서드)의 길이가 10라인을 넘어가지 않도록 구현한다.
> * else 예약어를 쓰지 않는다.
> 
> 질문)
> 
> > 선택한 점수의 이유를 적어주세요.
> > getter 및 setter를 사용했고, edit과 같이 메서드나 상수 이름에 축약형을 사용했기 때문입니다.
> 
> setter를 사용하셨다고 했는데, 어디에 있나요??
> 
> 남겨주신 질문에 답변 드려요~
> 
> > 1. TDD 방식으로 설계 시, 클래스 구현에 대한 고민
> 
> 사람마다 스타일이 다르겠지만, 저는 후자의 방식을 선호해요. 요구사항에 기반하여 대략적인 설계를 먼저 구상하고서 접근하다보니, 어떠한 객체들이 필요하고 어떠한 책임과 역할을 부여하면 좋을지 판단이 되거든요. 이렇게 기초를 쌓은 상태에서 테스트코드와 함께 구현에 대한 즉각적인 피드백을 주고받으며 개발하곤 합니다.
> 
> 제프가 생각하는 테스트코드의 장점과 역할은 무엇인가요?
> 
> > 2. 수정 기능이 필요한 객체 리스트를 특정 객체의 내부 필드로 사용하는 것과 일급컬렉션으로 직성하여 사용하는 것에 대한 고민
> 
> List에 행위를 부여하여 응집력을 높이고 역할을 부여할 수 있다면, 충분히 Attendances를 만들 이유가 될텐데요. 제프는 왜 일급컬렉션이 불변을 보장해야 한다고 보시나요??
> 
> > 3. Day 객체의 필요성에 대한 고민과 공휴일 처리에 대한 고민
> 
> 공휴일을 담지 않는, 즉 "주중"의 의미를 지닌 객체로 역할을 부여하면 어떨까 싶어요. 공휴일로 객체를 생성할 수 없도록 하면 되지 않을까요? 실제로 현재 `Day#checkHoliday` 메서드가 공휴일 검증에만 사용되는 것을 볼 수 있고, 요구사항도 공휴일에 대한 제약이 있다보니 Day 객체의 역할을 구체적으로 바꾸어봐도 좋아보여요.
> 
> > 4. 생성자 내부에 특정 로직을 넣어도 괜찮을까요
> 
> 생성자에 검증 로직을 두는 것은 옳다고 생각해요. 해당 객체의 생성에 대한 계약 조건이므로, 생성 시점에 validate하는 것은 지극히 당연하다고 봐요. 다만 로직에 따라 값을 달리 부여하거나 하는 행위는 음 정적 팩토리 메서드를 활용하는데요. 정적 팩토리 메서드에 대해 알아보시고, 생성자와 어떠한 차이가 있는지 그리고 제프는 어떠한 기준으로 생성자와 정적 팩토리 메서드를 나누어 사용할지 한번 고민해보시면 좋겠어요

horizonpioneer committed:
refactor: 파일 로드 시, 첫 줄을 건너뛰는 기능에 대해 의미 부여

horizonpioneer committed:
refactor: Crews에 크루 이름을 확인하고 등록하는 역할을 위임

horizonpioneer committed:
refactor: 매직넘버 제거

horizonpioneer committed:
refactor: Converter 내 중복 코드를 상수로 변환

horizonpioneer committed:
refactor: Crew 내부 메서드의 depth를 1로 수정

horizonpioneer commented:
> setter를 사용하셨다고 했는데, 어디에 있나요??

=>  Attendance 클래스 내부에서 출석 시간 변경을 담당하는 updateAttendanceTime 메서드입니다!


> 제프가 생각하는 테스트코드의 장점과 역할은 무엇인가요?

=> 테스트코드를 통해 코드에 대한 빠른 피드백을 받을 수 있다는 것이 장점인 것 같습니다. 그래서 개발 과정에 과감한 리팩토링이 가능하고 예상치 못한 문제를 사전에 방지할 수 있습니다. 테스트 코드는 결국 코드 변경에 따른 영향을 빠르게 파악할 수 있도록 도와주는 역할을 수행한다고 생각합니다.


> List에 행위를 부여하여 응집력을 높이고 역할을 부여할 수 있다면, 충분히 Attendances를 만들 이유가 될텐데요. 제프는 왜 일급컬렉션이 불변을 보장해야 한다고 보시나요??

=>  일급 컬렉션에 대한 자료를 찾아보다보니 불변을 보장한다는 이야기가 많더라구요! 개발 언어도 사회성을 띠고 있는 것 같아, 오해의 소지를 줘선 안된다고 생각했습니다 리스트에 행위만 부여할 수 있다면 해당 필드를 변경가능하도록 열어놔도 문제가 없을까요?

horizonpioneer committed:
fix: 리팩터링 후 출석 수정 과정에서 잘못된 문구가 출력되는 문제가 발생하여 재수정

horizonpioneer committed:
refactor: Crews 내부 메서드의 depth를 1로 수정

horizonpioneer committed:
style: stream 형식 줄바꿈

horizonpioneer committed:
feat: 분기 처리를 위해 역할에 맞는 Command 클래스 생성

horizonpioneer committed:
feat: 옵션에 따라 특정 Command 클래스를 생성하게 하는 CommandHandler 추가

horizonpioneer committed:
refactor: Controller 내부 분기 처리를 다형성을 활용한 방식으로 변경

horizonpioneer committed:
refactor: 제적 위험자의 정렬 로직을 OutputView로 이동

- CrewDtos 삭제
- 필터링된 CrewDto 리스트를 OutputView로 직접 넘기도록 수정

horizonpioneer committed:
refactor: isLate, isAbsent 조건 추가 및 메서드 순서 수정

- 출석 기록 조회 일자와 attendances.csv의 마지막 출석 기록 일자 사이는 결석 처리되기 때문에 그에 대한 조건문 추가
- 메서드를 public, private 순서대로 배치

horizonpioneer committed:
refactor: 지각횟수와 결석횟수를 계산하는 로직 내부 필터링 수정

horizonpioneer committed:
refactor: Attendance 내부 isLate, isAbsent 메서드를 public으로 변경

sinb57 commented on src/main/java/domain/command/CommandHandler.java:
```diff
@@ -0,0 +1,30 @@
+
+    private final Map<String, Command> commands = new HashMap<>();
+
+    public CommandHandler(Crews crews, InputView inputView, OutputView outputView) {
+        commands.put("1", new AttendanceCheckCommand(crews, inputView, outputView));
+        commands.put("2", new AttendanceUpdateCommand(crews, inputView, outputView));
+        commands.put("3", new AttendanceHistoryCommand(crews, inputView, outputView));
+        commands.put("4", new PenaltyCheckCommand(crews, outputView));
+    }
```
잘해주셨어요! 다만 InputView와 OutputView가 domain으로 흘러들어가네요ㅠ domain에서 view를 의존해도 좋을까요? domain과 view를 패키지로 나눈 이유가 무엇인가요?? 비즈니스와 UI 로직을 나누면 좋을 이유가 무엇일까요?

horizonpioneer replied:
도메인은 뷰의 영향을 받으면 안될 것 같습니다... 뷰의 변경에 따라 도메인 로직에 영향을 줄 수 있기 때문입니다. 분기 처리에 대해 방법을 찾다가 제일 중요한 부분을 놓쳤네요.
각 Command마다 필요한 요소와 반환값이 달라서 interface를 활용하기에도 쉽지 않아 보이는데 다른 방법이 있을까요?
제네릭을 활용해보려고 해도 타입 처리를 어떻게 해야할지 모르겠습니다.

horizonpioneer replied:
해당 부분 enum을 사용해 분기 처리하도록 변경했습니다!

sinb57 commented on src/main/java/domain/Day.java:
```diff
@@ -0,0 +1,59 @@
+
+    public Boolean checkHoliday() {
+        return getDayOfWeek().equals(DayOfWeek.SATURDAY) || getDayOfWeek().equals(DayOfWeek.SUNDAY);
+    }
```
2월 한정으로 정책을 수정했기에 쉬는날이 주말밖에 없지만, 본래의 기획대로 12월이 대상이었다면 성탄절이 조건에 추가되었어야 합니다. 이와 같이 월마다 각기 다른 쉬는날을 추가한다면 변경 범위가 어떻게 되나요??

horizonpioneer replied:
Holiday를 추가했고 해당 사항도 비교 항목으로 넣었습니다!
csv도 12월 출석부로 다시 만들어 파일 참조 링크를 변경했고, 오늘 날짜를 상수로 관리하도록 해서 직접 지정할 수 있도록 했습니다.

sinb57 commented on src/main/java/domain/DayOfWeek.java:
```diff
@@ -0,0 +1,45 @@
+
+    public static DayOfWeek getInstance(LocalDate date) {
+        return Arrays.stream(values())
+                .filter(value -> value.id.equals(date.getDayOfWeek().getValue()))
```
getValue() 메서드를 들여다보면 알겠지만, ordinal()에 의존하기 때문에 지양해야 합니다. 여유 되실 때, 이펙티브 자바의 아이템 35를 읽어보세요.

horizonpioneer replied:
감사합니다! 해당 부분을 찾아 읽어보니 순서가 변경됨에 따라 값이 달라질 수 있기 때문에 ordinal에 의존하는 것을 지양해야 한다는 것을 알게 되었습니다. 현재 DayOfWeek를 id 대신 필드로 두어 비교하는 방식으로 변경해보니 기존 방법보다 안정성면에서 더 나은 코드라는 생각도 들었습니다.
근데 혹시 DayOfWeek 같은 일주일이 정해진 enum을 사용하는 부분에 있어서도 사용을 피하는게 맞을까요...? 이프의 의견이 궁금해 여쭤봅니다!

sinb57 replied:
> DayOfWeek 같은 일주일이 정해진 enum을 사용하는 부분에 있어서도 사용을 피하는게 맞을까요...?

이해하지 못했어요.. 어떠한 사용을 피하나요?

horizonpioneer replied:
ordinal은 결국 열거된 상수의 순서에 영향을 받는 것인데 DayOfWeek 같이 월,화,수,목,금,토,일로 만연하게 정해진 부분에 대해서도 변경 가능성을 염두에 둬야 할지에 대한 질문이었습니다!

sinb57 replied:
아하! 이해했습니다~ java.time.DayOfWeek는 라이브러리로 제공된 Enum이기 때문에 순서가 변경될 가능성은 몹시 낮아요! 따라서 변경 가능성은 없다고 봐도 좋지만, 일반적으로 enum의 순서에 대한 ordinal의 활용은 안티패턴이기 때문에 지양하면 좋습니다.
그리고 getValue()를 통해 enum의 순서를 가져오는 것은 단순히 int 값을 반환하는 것이기 때문에, 의미를 알기 어려운 변수가 될 수 있어요.

horizonpioneer committed:
docs: 구현 기능에 대한 체크

sinb57 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public Attendance findByDate(Integer dayOfMonth) {
+        LocalDate today = LocalDate.now();
```
LocalDate를 Day로 포장했음에도 활용하지 않는 이유가 궁금해요.

horizonpioneer replied:
findByDate의 인자를 Day로 두어 컴파일 단계에서 다른 값이 들어왔을 때 미리 알 수 있도록 하는 방식을 말씀하시는 걸까요?
컨트롤러에서 inputView를 통해 일자만 입력받기 때문에, findByDate 내부에서 해당 일자를 통해 LocalDate를 만들면서 validateDayOfMonth 검증을 하기 위해 해당 방식처럼 구현했는데 혹시 다른 부분에 활용할 방법이 있을까요?

sinb57 replied:
아 `LocalDate today = LocalDate.now();`에 대한 질문이었어요. LocalDate를 Day로 포장했는데, Day가 아닌 LocalDate를 여러곳에서 사용하더라고요. Day와는 다른 목적으로 사용하는 걸까요? Day는 어떻게 다른가요?

horizonpioneer replied:
아하! 기존의 Day는 공휴일 처리 및 출석시간에 따라 해당 날짜의 출석 상태를 체크하기 위해 사용했었는데 Day에 역할을 추가해서 Day를 활용하도록 변경했습니다. 찾아보니 날짜 함수를 사용해서 더 간단하게 표현할 방법도 있고 Day 책임으로 분리하는 것이 유지 보수에 도움이 될 것 같네요.

sinb57 commented on src/main/java/config/AttendanceHistoryLoader.java:
```diff
@@ -0,0 +1,59 @@
+        Crews crews = new Crews(new ArrayList<>());
+
+        try (BufferedReader reader = new BufferedReader(new FileReader("src/main/resources/attendances.csv"))) {
+            skipItemTitle(reader);
```
의미 부여를 해주셨네요👍 네이밍 중 `Item`은 무엇을 의미하는지 궁금해요.

horizonpioneer replied:
'항목' 이라는 의미로 사용했습니다!

sinb57 replied:
추상적인 용어는 지양해주는게 좋아요. 저는 '요소'라고 이해했거든요. `skipTitle`이라고 해도 좋지 않을까 싶고, 저라면 `ignoreTitleOfFile`로 할 것 같아요.

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,32 @@
+        String option = "";
+        CommandHandler commandHandler = new CommandHandler(crews, inputView, outputView);
+
+        while (!option.equals("Q")) {
```
매직 넘버가 보이네요

sinb57 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,90 @@
+
+        today.withDayOfMonth(1)
+                .datesUntil(today)
+                .filter(date -> date.getDayOfWeek().getValue() != SATURDAY && date.getDayOfWeek().getValue() != SUNDAY)
```
SATURDAY와 SUNDAY의 선언부를 봤을 때, 숫자 값이어서 의아했어요. 로직의 getValue()를 보고서야 이해했네요. enum의 ordinary에 의존하지 말고 아래와 같이 enum 값 자체에 의존하면 어떨까요?
`date.getDayOfWeek() != java.time.DayOfWeek.SATURDAY`

horizonpioneer replied:
enum을 자체적으로 비교하는게 더 쉽게 코드를 파악할 수 있겠네요! 감사합니다

sinb57 reviewed (changes requested):
안녕하세요 제프! 피드백 반영하느라 고생 많으셔요~
추가로 남긴 코멘트와 더불어 기존 피드백도 같이 반영해주세요

---

> 2024년 12월 한 달 동안 시범적으로 최소한의 기능을 갖춘 출석 시스템을 개발하여 출석을 체계적으로 관리하기로 하였습니다

요구사항을 보면, 현재는 12월 한정 시범 운영이지만 시스템이 확장될 가능성을 엿볼 수 있어요. 이 경우 현재의 구현에서 확장에 대응하려면 어떠한 변경이 있을지 생각해보시겠어요? 과연 현재의 코드는 변경에 열려 있나요?

---

> setter를 사용하셨다고 했는데, 어디에 있나요??
> => Attendance 클래스 내부에서 출석 시간 변경을 담당하는 updateAttendanceTime 메서드입니다!

아하! updateAttendanceTime는 setter로 봐야 할까요? 제프는 setter를 무엇이라고 보시나요? 왜 많은 사람들이 setter를 지양해야 한다고 할까요?

> 제프가 생각하는 테스트코드의 장점과 역할은 무엇인가요?
> => 테스트코드를 통해 코드에 대한 빠른 피드백을 받을 수 있다는 것이 장점인 것 같습니다. 그래서 개발 과정에 과감한 리팩토링이 가능하고 예상치 못한 문제를 사전에 방지할 수 있습니다. 테스트 코드는 결국 코드 변경에 따른 영향을 빠르게 파악할 수 있도록 도와주는 역할을 수행한다고 생각합니다.

좋은 답변 감사해요~ 답변을 보니, 제프는 리팩터링을 위한 수단으로 테스트코드를 바라보고 있는 듯 해요! "피드백을 받을 수 있다고" 하셨는데, 제프가 생각하는 피드백은 무엇인가요? 실행 결과 성공하고 실패하는 것만을 의미하는 것일까요?

> List에 행위를 부여하여 응집력을 높이고 역할을 부여할 수 있다면, 충분히 Attendances를 만들 이유가 될텐데요. 제프는 왜 일급컬렉션이 불변을 보장해야 한다고 보시나요??
> => 일급 컬렉션에 대한 자료를 찾아보다보니 불변을 보장한다는 이야기가 많더라구요! 개발 언어도 사회성을 띠고 있는 것 같아, 오해의 소지를 줘선 안된다고 생각했습니다 리스트에 행위만 부여할 수 있다면 해당 필드를 변경가능하도록 열어놔도 문제가 없을까요?

"개발 언어도 사회성을 띠고 있는 것 같아, 오해의 소지를 줘선 안된다"에 동의하지만, "일급 컬렉션에 대한 자료를 찾아보다보니 불변을 보장한다는 이야기가 많더라구요"에 동의는 어렵네요. 어떠한 개념에 대해 받아들일 때는 등장한 이유와 배경을 이해하는 것이 중요하고 어떠한 성질을 지니고 있는지 등 원초적인 조사가 필요하다고 생각해요.
일급 컬렉션에서 "일급"은 무슨 의미일까요? 어찌보면 이 때문에 불변성을 보장해야 한다는 내용이 탄생했을지도 몰라요. 일급 컬렉션을 처음 다루는 `The ThoughtWorks Anthology(소트웍스 엔솔로지)`을 찾아보시는 것도 좋겠어요.
저는 해당 케이스의 경우 필드를 변경가능토록 열어놔도 좋다고 생각해요. 만약 불변을 무조건 보장해야 한다면, Crews 객체를 다시 살펴보길 바래요. 불변성을 보장하는 것은 무슨 의미일까요? Crews는 불변을 보장하고 있나요?

horizonpioneer committed:
refactor: 매직 넘버 제거

horizonpioneer committed:
refactor: DayOfWeek의 value를 비교하는 것이 아닌 enum 자체를 비교하도록 수정

horizonpioneer committed:
refactor: 오늘 날짜를 직접 지정할 수 있도록 상수 처리

horizonpioneer committed:
refactor: Holiday enum 추가

horizonpioneer committed:
refactor: AttendanceHistoryLoader를 도메인 패키지로 이동

horizonpioneer committed:
refactor: 12월 출석부 추가 및 12월 기준으로 파일 링크 변경

horizonpioneer committed:
refactor: 메인 로직의 분기 처리를 enum을 활용한 방식으로 변경

horizonpioneer committed:
refactor: DayOfWeek -> AttendanceStandard로 변경

- id 필드를 DayOfWeek 타입으로 수정
- 내부 메서드에서 ordinal에 의존한 필터링이 아닌 DayOfWeek를 비교하는 필터링으로 수정

horizonpioneer commented:
> 아하! updateAttendanceTime는 setter로 봐야 할까요? 제프는 setter를 무엇이라고 보시나요? 왜 많은 사람들이 setter를 지양해야 한다고 할까요?

=> 다시 살펴보니 여기선 updateStatus()가 있으므로 setter라고 보긴 어려운 것 같습니다. 단순히 필드를 변경하는 역할만 수행하는게 아니라 출석 상태까지 업데이트하기 때문입니다. 값 설정과 관련된 비즈니스 로직을 포함한 메서드로 보는 것이 더 적합할 것 같습니다.
setter를 지양해야 하는 이유는 값을 변경하는 의도를 파악하기 어렵기 때문인 점이 가장 큰 것 같습니다.

> 좋은 답변 감사해요~ 답변을 보니, 제프는 리팩터링을 위한 수단으로 테스트코드를 바라보고 있는 듯 해요! "피드백을 받을 수 있다고" 하셨는데, 제프가 생각하는 피드백은 무엇인가요? 실행 결과 성공하고 실패하는 것만을 의미하는 것일까요?

=>  사실 피드백 자체는 작동여부와 코드 품질에 대해 다시 한 번 돌아보게 하는 장치라고 생각했습니다. 이프의 질문을 듣고 다시 한 번 피드백에 대해, 테스트 코드의 의미에 대해 생각해보았습니다. 테스트를 구현하면서 기존의 코드를 되돌아본 적이 있는데, 그 때마다 더 간결하게 혹은 객체간 결합도를 줄였어야 했다고 생각하면서 프로덕션 코드 품질에 대해 다시 한 번 생각하게 되는 계기가 되었습니다. 따라서 코드 품질을 개선할 수 있다는 점도 테스트 코드 작성의 장점 중 하나라고 생각합니다.

> "개발 언어도 사회성을 띠고 있는 것 같아, 오해의 소지를 줘선 안된다"에 동의하지만, "일급 컬렉션에 대한 자료를 찾아보다보니 불변을 보장한다는 이야기가 많더라구요"에 동의는 어렵네요. 어떠한 개념에 대해 받아들일 때는 등장한 이유와 배경을 이해하는 것이 중요하고 어떠한 성질을 지니고 있는지 등 원초적인 조사가 필요하다고 생각해요. 일급 컬렉션에서 "일급"은 무슨 의미일까요? 어찌보면 이 때문에 불변성을 보장해야 한다는 내용이 탄생했을지도 몰라요. 일급 컬렉션을 처음 다루는 `The ThoughtWorks Anthology(소트웍스 엔솔로지)`을 찾아보시는 것도 좋겠어요. 저는 해당 케이스의 경우 필드를 변경가능토록 열어놔도 좋다고 생각해요. 만약 불변을 무조건 보장해야 한다면, Crews 객체를 다시 살펴보길 바래요. 불변성을 보장하는 것은 무슨 의미일까요? Crews는 불변을 보장하고 있나요?

=> 일급이라는 용어에 대해 찾아보니 자체적으로 유효성 검사, 비즈니스 로직, 또는 그 컬렉션에 관련된 기능을 처리할 수 있는 객체를 의미하는 것 같습니다. 일반적으로 객체에서 불변성을 권장하는 정도고 반드시 막아야 한다는 내용은 없다보니, 일급 컬렉션도 마찬가지로 불변성을 최대한 권장한다는 의미로 보는게 맞는 것 같습니다. Crews 객체도 내부 로직 중, add를 통해 crew를 넣는 부분이 있어 불변을 보장한다고 말하기 어려운 것 같습니다... 제가 이해한 부분이 맞을까요?

sinb57 reviewed (changes requested):
안녕하세요 제프~ 피드백 반영해주셔서 감사해요! 논의가 필요한 부분이 있어서 따로 추가 피드백은 없었어요~ 코멘트 확인하시고 답변 부탁드려요!

---

> 일급이라는 용어에 대해 찾아보니 자체적으로 유효성 검사, 비즈니스 로직, 또는 그 컬렉션에 관련된 기능을 처리할 수 있는 객체를 의미하는 것 같습니다. 일반적으로 객체에서 불변성을 권장하는 정도고 반드시 막아야 한다는 내용은 없다보니, 일급 컬렉션도 마찬가지로 불변성을 최대한 권장한다는 의미로 보는게 맞는 것 같습니다. Crews 객체도 내부 로직 중, add를 통해 crew를 넣는 부분이 있어 불변을 보장한다고 말하기 어려운 것 같습니다... 제가 이해한 부분이 맞을까요?

어려운 질문에도 답변주셔서 감사해요. 답변 주신 바와 같이 일급 컬렉션은 불변을 무조건 지켜야 하는 것은 아니예요. 여기서 불변을 보장한다는 것은 다양한 견해가 있겠지만, 완전한 불변은 내부의 요소도 불변인 상태를 의미해요. 제가 Crews에 대해 질문 드렸던 이유는, 불변을 보장한다는 이유로 Attendances 객체의 존재를 반대하셨지만 Crews는 불변을 보장하고 있지 않았기 때문인데요. 지금 Crews를 보면 List<Crew>의 요소 하나하나에 대해 변경을 막지 않았어요. 때문에 이는 완전한 불변이라고 보기 어렵다고 생각했어요. 불변에 대해서는 "방어적 복사"를 찾아보시는 것도 큰 도움 될거예요!

horizonpioneer committed:
refactor: skipItemTitle -> ignoreTitleOfFile로 메서드명 수정

horizonpioneer committed:
refactor: InputView에서 Converter를 호출해 변환값을 넘기도록 수정

horizonpioneer committed:
refactor: 재사용되지 않는 상수 제거 및 문자열 그대로 출력문에 삽입

horizonpioneer committed:
refactor: Day를 활용하여 LocalDate 데이터를 다루도록 수정

horizonpioneer committed:
refactor: 제적 위험자가 아닐 경우, 대상자라는 문구가 나오지 않도록 수정

horizonpioneer committed:
refactor: StandardDate의 상수 타입을 Day로 변경

horizonpioneer committed:
refactor: 날짜를 다루는 로직을 Day로 이동 및 추가

horizonpioneer committed:
refactor: 도메인에서 Dto를 의존하지 않도록 수정

- 도메인 내부에서 toDto 메서드 삭제
- AttendanceDto의 경우 Converter에서 변환하도록 수정
- crewDto의 경우, Controller의 필드인 crews를 사용해 CrewDto 리스트로 변환하도록 수정
- 출력부에서 depth가 2인 메서드 분리

horizonpioneer commented:
> 어려운 질문에도 답변주셔서 감사해요. 답변 주신 바와 같이 일급 컬렉션은 불변을 무조건 지켜야 하는 것은 아니예요. 여기서 불변을 보장한다는 것은 다양한 견해가 있겠지만, 완전한 불변은 내부의 요소도 불변인 상태를 의미해요. 제가 Crews에 대해 질문 드렸던 이유는, 불변을 보장한다는 이유로 Attendances 객체의 존재를 반대하셨지만 Crews는 불변을 보장하고 있지 않았기 때문인데요. 지금 Crews를 보면 List의 요소 하나하나에 대해 변경을 막지 않았어요. 때문에 이는 완전한 불변이라고 보기 어렵다고 생각했어요. 불변에 대해서는 "방어적 복사"를 찾아보시는 것도 큰 도움 될거예요!

=> 불변성 문제로 일급 컬렉션 사용이 고민된다고 말씀드렸었는데, Crews에는 특정 불변 처리를 하지 않았다는 모순이 있었네요.
Crew가 가지고 있는 Attendance 리스트를 수정 가능하게 구현해야 해서 Crews에 별 다른 처리를 하지 않았던 것인데 코드 자체에 답이 있었네요! 감사합니다

sinb57 commented on README.md:
```diff
@@ -1,3 +1,140 @@
+
+# 입력
+
+- [x]  실행할 기능을 숫자로 입력 받는다
```
d

sinb57 commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,58 @@
+public class Attendance {
+    private final Day day;
+    private Boolean isLate = false;
+    private Boolean isAbsent = false;
```
사용하지 않는 필드가 있네요

sinb57 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,86 @@
+                .orElseThrow(() -> new IllegalArgumentException("[ERROR] 해당 날짜는 출석일이 아닙니다."));
+    }
+
+    public List<Attendance> getAttendances() {
+        attendances.sort(Comparator.comparing((Attendance attendance) -> attendance.getDate()));
+        return List.copyOf(attendances);
+    }
```
"정렬" 행위를 도메인에서 행해야 하는지 의문이에요.

horizonpioneer replied:
정렬은 출력뷰쪽에서 담당하는게 맞는 것 같습니다!

sinb57 commented on src/main/java/util/Converter.java:
```diff
@@ -0,0 +1,49 @@
+import java.time.LocalTime;
+import java.time.format.DateTimeFormatter;
+
+public class Converter {
```
"변환"에 해당하는 행위가 모여있군요! 프로젝트가 커질 수록 많은 "변환"이 이 곳에 담길 텐데, 해당 클래스가 무거워져도 괜찮나요?

sinb57 commented on src/main/java/domain/constant/StandardDate.java:
```diff
@@ -0,0 +1,13 @@
+
+public class StandardDate {
+
+    public static final Day TODAY = new Day(LocalDate.of(2024, 12, 26));
```
오늘의 날짜가 아니라, 2024.12.26로 고정한 이유가 있을까요? 프로그램이 해당 날짜로 고정되어 동일한 동작을 할텐데, 기획의 의도와 맞을까요?

horizonpioneer replied:
크리스마스 공휴일을 확인할 수 있도록 임시로 오늘 일자를 크리스마스 이후 날짜로 설정했습니다. 수정 일자만 받아서 수정 기능을 구현해야 하는 것으로 봤을 때 해당 달의 출석으로 한정지어야 한다고 생각했습니다.

horizonpioneer replied:
크리스마스 공휴일을 확인할 수 있도록 임시로 오늘 일자를 크리스마스 이후 날짜로 설정했습니다. 수정 일자만 받아서 수정 기능을 구현해야 하는 것으로 봤을 때 해당 달의 출석으로 한정지어야 한다고 생각했습니다.
사실 기획의 의도대로 구현한다고 하면 하나의 달에 한정된 프로젝트가 될 것이고, LocalDate.now()로 설정해야 한다고 생각합니다. 다만 주어진 데이터가 12월 데이터이고, 공휴일도 크리스마스로 설정되어있어서 LocalDate.now()로 설정하게 되면 2월로 설정되어 원하는 값을 얻기 어려울 것 같습니다.
오늘 일자를 입력 받아 해당 달의 데이터로만 한정짓게 하거나, 수정 기능에 연, 월, 일을 모두 입력 받도록 해서 기능을 확대해야 할 것 같은데 어떤 방향으로 설정하는게 좋을까요?

horizonpioneer replied:
크리스마스 공휴일을 확인할 수 있도록 임시로 오늘 일자를 크리스마스 이후 날짜로 설정했습니다. 수정 일자만 받아서 수정 기능을 구현해야 하는 것으로 봤을 때 해당 달의 출석으로 한정지어야 한다고 생각했습니다.
사실 기획의 의도대로 구현한다고 하면 하나의 달에 한정된 프로젝트가 될 것이고, LocalDate.now()로 설정해야 한다고 생각합니다. 다만 주어진 데이터가 12월 데이터이고, 공휴일도 크리스마스로 설정되어있어서 LocalDate.now()로 설정하게 되면 2월로 설정되어 원하는 값을 얻기 어려울 것 같습니다.
오늘 일자를 입력 받아 해당 달의 데이터로만 한정하거나, 수정 기능에 연, 월, 일을 모두 입력 받도록 해서 기능을 확대해야 할 것 같은데 어떤 방향으로 설정하는게 좋을까요?

horizonpioneer replied:
크리스마스 공휴일을 확인할 수 있도록 임시로 오늘 일자를 크리스마스 이후 날짜로 설정했습니다. 수정 일자만 받아서 수정 기능을 구현해야 하는 것으로 봤을 때 해당 달의 출석으로 한정지어야 한다고 생각했습니다.
사실 기획의 의도대로 구현한다고 하면 하나의 달에 한정된 프로젝트가 될 것이고, LocalDate.now()로 설정해야 한다고 생각합니다. 다만 주어진 데이터가 12월 데이터이고, 공휴일도 크리스마스로 설정되어있어서 LocalDate.now()로 설정하게 되면 2월로 설정되어 원하는 값을 얻기 어려울 것 같습니다.
오늘 일자를 입력 받아 해당 달의 데이터로만 한정하거나 수정 기능에 연, 월, 일을 모두 입력 받도록 해서 기능을 확대해야 할 것 같은데 어떤 방향으로 설정하는게 좋을까요?

sinb57 replied:
제프가 생각하는 방향으로 기획을 확장/수정해서 진행해보면 좋겠어요ㅎ 정답은 없어요~!! 어떠한 방향이 사용자에게 불편하지 않으면서 기획의 의도를 잘 충족할 수 있을지, 시간 되신다면 이러한 고민도 가져보세요~ 본래는 기획자의 역할이지만, 개발자여도 자유로이 생각할 수 있어야 한다고 봐요.

sinb57 reviewed (changes requested):
안녕하세요 제프~ 고생 많으셨어요!! 추가 코멘트랑 더불어 resolve 처리안한 코멘트들도 같이 확인해주세요~

horizonpioneer committed:
refactor: equals 앞뒤 순서 변경을 통해 NullPointerException 방지

horizonpioneer committed:
refactor: 도메인 내부 정렬 로직을 OutputView로 이동

horizonpioneer committed:
refactor: 모든 도메인 Dto 제거 및 Converter 역할을 View로 이동

horizonpioneer committed:
refactor: loadCrews의 인자로 FileReader를 받도록 수정

horizonpioneer committed:
feat: AttendanceHistoryLoaderTest 추가

sinb57 reviewed (approved):
안녕하세요 제프~ 1단계 진행하느라 고생 많으셨어요!
긴 시간동안 많은 리뷰를 드렸는데, 꼼꼼히 반영해주셔서 감사해요!!
2단계 진행을 위해 머지하겠습니다! 2단계도 응원할게요~
